Fix invoking of `FileSystemWatcher`

thomsj opened the pull request:
Call `protected` `On...` event raising methods from `Notify...` methods,
to invoke `SynchronizingObject` when required.

Fix #52644

thomsj committed:
Fix invoking of `FileSystemWatcher`

Call `protected` `On...` event raising methods from `Notify...` methods,
to invoke `SynchronizingObject` when required.

Fix #52644

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=53151) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=53151) before we can accept your contribution.<br/><br/>:x: thomsj [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=53151)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=53151) it.</sub>

thomsj commented on src/libraries/System.IO.FileSystem.Watcher/tests/Utility/TestFileSystemWatcher.cs:
```diff
@@ -40,5 +45,35 @@ public void CallOnRenamed(RenamedEventArgs e)
+                new Type[] { typeof(WatcherChangeTypes), typeof(ReadOnlySpan<char>) },
+                null);
+
+            notifyFileSystemEventArgs.CreateDelegate<NotifyFileSystemEventArgs>(this)(changeType, name);
```
Would you prefer [`CreateDelegate(Type, Object)`](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.methodinfo.createdelegate?view=net-5.0#System_Reflection_MethodInfo_CreateDelegate_System_Type_System_Object_), so this can be backported?

stephentoub commented on src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs:
```diff
@@ -447,11 +446,11 @@ private void NotifyRenameEventArgs(WatcherChangeTypes action, ReadOnlySpan<char>
+            if (raiser != null && MatchPattern(name.IsEmpty ? _directory : name))
             {
-                handler(this, new FileSystemEventArgs(changeType, _directory, name.IsEmpty ? null : name.ToString()));
+                raiser(new FileSystemEventArgs(changeType, _directory, name.IsEmpty ? null : name.ToString()));
```
These changes will result in delegates being allocated for every event raised, which is not something we want to do.  It's also unnecessary.

This method can be reverted (along with the changes to GetHandler), and then this one line changed to just be:
```C#
InvokeOn(new FileSystemEventArgs(changeType, _directory, name.IsEmpty ? null : name.ToString()), handler);
```

thomsj replied:
I have made the change you suggested and reverted, except I have removed the `handler != null` check, since the check is also in `InvokeOn()`.

thomsj replied:
Reverting means that `OnCreated()`, `OnChanged()` and `OnDeleted()` are not used by `FileSystemWatcher`. Would you prefer it if I called `CallNotifyFileSystemEventArgs()` instead of these methods, in the unit tests such as this:

https://github.com/dotnet/runtime/blob/fda04b9d36c273e3fcef1fe911d8502de944125a/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.cs#L324-L345

stephentoub replied:
> I have made the change you suggested and reverted, except I have removed the handler != null check, since the check is also in InvokeOn().

Please put back the null checks.  We don't want to do work like allocating the event args or performing a name match if there's no handler registered.  The extra null check won't matter.

stephentoub replied:
> Would you prefer it if I called CallNotifyFileSystemEventArgs() instead of these methods, in the unit tests such as this:

No, these are tests of the protected methods; they're part of the exposed surface area and we need to continue to validate them even if we don't use them ourselves.

stephentoub commented on src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs:
```diff
@@ -460,11 +459,11 @@ private void NotifyFileSystemEventArgs(WatcherChangeTypes changeType, ReadOnlySp
         private void NotifyFileSystemEventArgs(WatcherChangeTypes changeType, string name)
         {
-            FileSystemEventHandler? handler = GetHandler(changeType);
+            Action<FileSystemEventArgs>? raiser = GetRaiser(changeType);
 
-            if (handler != null && MatchPattern(string.IsNullOrEmpty(name) ? _directory : name))
+            if (raiser != null && MatchPattern(string.IsNullOrEmpty(name) ? _directory : name))
             {
-                handler(this, new FileSystemEventArgs(changeType, _directory, name));
+                raiser(new FileSystemEventArgs(changeType, _directory, name));
             }
         }
```
You're correct that it's not being used right now.  However, I think that's arguably a perf bug.  Call sites that already have a string (but stored in a span) are then calling the span-based overload, and then having to call ToString on the span to produce a new one.  It'd be better if those call sites, if they have a string, would actually use the string-based overload, or potentially they'd be consolidated into a single method that accepted both a span and an optional string, using the string instead of ToString if the string is non-null.

That's not a change for this PR, but it's to say let's not remove this yet.

thomsj commented on src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs:
```diff
@@ -408,7 +408,7 @@ private bool MatchPattern(ReadOnlySpan<char> relativePath)
         /// </summary>
         private void NotifyInternalBufferOverflowEvent()
         {
-            _onErrorHandler?.Invoke(this, new ErrorEventArgs(
+            OnError(new ErrorEventArgs(
                     new InternalBufferOverflowException(SR.Format(SR.FSW_BufferOverflow, _directory))));
         }
```
Do you want a null check of `_onErrorHandler` before calling `OnError()`?

stephentoub replied:
Yup, thanks.

thomsj commented on src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs:
```diff
@@ -417,12 +417,10 @@ private void NotifyInternalBufferOverflowEvent()
-            // filter if there's no handler or neither new name or old name match a specified pattern
-            RenamedEventHandler? handler = _onRenamedHandler;
-            if (handler != null &&
-                (MatchPattern(name) || MatchPattern(oldName)))
+            // filter if neither new name or old name match a specified pattern
+            if (MatchPattern(name) || MatchPattern(oldName))
             {
-                handler(this, new RenamedEventArgs(action, _directory, name.IsEmpty ? null : name.ToString(), oldName.IsEmpty ? null : oldName.ToString()));
+                OnRenamed(new RenamedEventArgs(action, _directory, name.IsEmpty ? null : name.ToString(), oldName.IsEmpty ? null : oldName.ToString()));
             }
         }
```
Null check `_onRenamedHandler` first?

```cs
// filter if there's no handler or neither new name or old name match a specified pattern
if (_onRenamedHandler != null &&
    (MatchPattern(name) || MatchPattern(oldName)))
{
    OnRenamed(new RenamedEventArgs(action, _directory, name.IsEmpty ? null : name.ToString(), oldName.IsEmpty ? null : oldName.ToString()));
}
```

stephentoub replied:
Yup.

thomsj reviewed (commented):
> > I have made the change you suggested and reverted, except I have removed the handler != null check, since the check is also in InvokeOn().
> 
> Please put back the null checks. We don't want to do work like allocating the event args or performing a name match if there's no handler registered. The extra null check won't matter.

stephentoub commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.cs:
```diff
@@ -102,24 +102,24 @@ public void SynchronizingObject_CalledOnEvent(WatcherChangeTypes expectedChangeT
                 {
                     watcher.Created += dele;
-                    watcher.CallOnCreated(new FileSystemEventArgs(WatcherChangeTypes.Created, "test", "name"));
+                    watcher.CallNotifyFileSystemEventArgs(WatcherChangeTypes.Created, "name");
```
These tests shouldn't be changing.  They were validating the behavior for when the OnXx methods were explicitly called, which, for example, a type derived from FileSystemWatcher could do (that's why these are all using a TestFileSystemWatcher, which is a derived type that just exposes those On methods for use).

So all of these tests changes should be reverted.  Instead, tests should be added that validate that actually making file changes which in turn trigger the FileSystemWatcher to invoke event handlers does so with those invocations in the right place.

stephentoub reviewed (commented):
Thanks for working on this.  The product changes now look good.  The tests need some further changes, as noted in my comments.  Let us know if you need any assistance.

thomsj commented:
Is there a reason why the tests in the `.Move.cs` files aren't prefixed with `FileSystemWatcher`?

https://github.com/dotnet/runtime/blob/86cebb31b1d216b1582f20352f16361ea67de5da/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Move.cs#L17-L20

The tests in the `.Changed.cs`, `.Create.cs`, `.Delete.cs` and `.NotifyFilter.cs` files are all prefixed. For example:

https://github.com/dotnet/runtime/blob/86cebb31b1d216b1582f20352f16361ea67de5da/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Changed.cs#L13-L24

thomsj commented:
These two tests are very similar:

https://github.com/dotnet/runtime/blob/86cebb31b1d216b1582f20352f16361ea67de5da/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Move.cs#L15-L20

https://github.com/dotnet/runtime/blob/86cebb31b1d216b1582f20352f16361ea67de5da/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Move.cs#L90-L94

Does it make sense that one is specific to Windows but the other isn't?

Sorry if I've missed something.

thomsj commented:
Should these two tests be combined?

https://github.com/dotnet/runtime/blob/86cebb31b1d216b1582f20352f16361ea67de5da/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Move.cs#L13-L25

thomsj commented:
Are the `TestISynchronizeInvoke` delegates, such as the following, supposed to be invoked?

https://github.com/dotnet/runtime/blob/6474c725025a8b29b4265c4d9485018982ee6b16/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.cs#L121-L137

If so, should [`TestISynchronizeInvoke.BeginInvoke()`](https://github.com/dotnet/runtime/blob/6474c725025a8b29b4265c4d9485018982ee6b16/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.cs#L60-L65) be changed to something like the below?

```cs
public IAsyncResult BeginInvoke(Delegate method, object[] args)
{
    Assert.Equal(ExpectedDelegate, method);
    BeginInvoke_Called = true;
    method.DynamicInvoke(args[0], args[1]);
    return null;
}
```

thomsj commented:
Is the `// Uses P/Invokes` comment here accurate?

https://github.com/dotnet/runtime/blob/02f70d0b903422282cd7ba8037de6b66ea0b7a2d/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.InternalBufferSize.cs#L41-L45

I tried `[PlatformSpecific(TestPlatforms.Windows | TestPlatforms.Linux)]` but failed to get the error on Linux.

Also, if the following comment is correct, why was the calculation changed?

https://github.com/dotnet/runtime/blob/febb25693e0547e8ec38aba66477b5d93830ed37/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.InternalBufferSize.cs#L35-L37

Previous:
https://github.com/dotnet/runtime/blob/c0f649508b0e1da68644cd32899f920b9b857cd4/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.InternalBufferSize.cs#L39

Current (introduced at same time as comment):
https://github.com/dotnet/runtime/blob/febb25693e0547e8ec38aba66477b5d93830ed37/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.InternalBufferSize.cs#L51

thomsj commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.InternalBufferSize.cs:
```diff
@@ -44,40 +44,80 @@ public class InternalBufferSizeTests : FileSystemWatcherTest
+        {
+            var watcher = new FileSystemWatcher(testDirectoryPath, Path.GetFileName(filePath));
+
+            // block the handling thread
+            watcher.Changed += (o, e) => unblockHandler.WaitOne();
```
Will this work with a recreated watcher?

stephentoub replied:
I don't understand the question.  Can you elaborate?

thomsj replied:
https://github.com/dotnet/runtime/blob/f044460da04f0833990325b144f8f26478293cca/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs#L387-L425

https://github.com/dotnet/runtime/blob/f044460da04f0833990325b144f8f26478293cca/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs#L511-L528

If `watcher` is created in `RecreateWatcher()`, will the correct thread be blocked?

stephentoub replied:
Ah, no.  The new watcher you're creating won't have the event handler that was originally stored into the first watcher.

thomsj replied:
I assume I should leave this as is, since this is a pre-existing flaw with `FileSystemWatcher_InternalBufferSize()`?

stephentoub replied:
Yeah, we can clean up stuff later if it's an issue.

stephentoub replied:
Except, is this actually pre-existing?  Where is the existing FileSystemWatcher_InternalBufferSize recreating the watcher and re-setting this handler?

thomsj replied:
In the `main` branch, `FileSystemWatcher_InternalBufferSize()` calls one of the test methods:

https://github.com/dotnet/runtime/blob/016851a183c66333f3087312b458452a90154726/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.InternalBufferSize.cs#L76-L79

Both of them call `TryErrorEvent()`:

https://github.com/dotnet/runtime/blob/016851a183c66333f3087312b458452a90154726/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs#L351-L355

https://github.com/dotnet/runtime/blob/016851a183c66333f3087312b458452a90154726/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs#L364-L368

`TryErrorEvent()` recreates the watcher, but the handler is never re-set:

https://github.com/dotnet/runtime/blob/016851a183c66333f3087312b458452a90154726/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs#L370-L431

thomsj commented on src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs:
```diff
@@ -374,25 +383,17 @@ public static void ExpectNoError(FileSystemWatcher watcher, Action action, Actio
-                        Filter = watcher.Filter,
-                        Path = watcher.Path,
-                        InternalBufferSize = watcher.InternalBufferSize
-                    };
+                    watcher = RecreateWatcher(watcher);
```
Should the recreated watchers be disposed?

stephentoub replied:
Ideally, but that's also beyond the scope of your fix.

Let's keep this PR constrained to the minimal changes necessary.

stephentoub commented:
@thomsj, all fine questions, but most of them don't seem related to this PR.  Are there specific questions related to this fix you still have?  Thanks.

stephentoub commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Move.cs:
```diff
@@ -93,6 +93,27 @@ public void Directory_Move_With_Set_NotifyFilter()
+        [Fact]
+        public void FileSystemWatcher_Directory_Move_SynchronizingObject()
+        {
+            TestISynchronizeInvoke invoker = new TestISynchronizeInvoke();
```
nit: for all of these tests, you can move this into the using block, just above where you store it into SynchronizingObject.

thomsj replied:
I've moved this for all the tests. I think I probably put it where I did and didn't use `var` because that's how it's done in `tests/FileSystemWatcher.cs`

stephentoub commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.InternalBufferSize.cs:
```diff
@@ -44,40 +44,80 @@ public class InternalBufferSizeTests : FileSystemWatcherTest
         }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]  // Uses P/Invokes
```
To answer your question, no, I don't believe this comment is still valid.  I'm guessing it's marked Windows-specific because it's relying on implementation details around Windows buffering, but that's just a guess.

thomsj replied:
I've removed this comment.

thomsj commented on src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs:
```diff
@@ -607,5 +602,25 @@ void AddEvent(WatcherChangeTypes eventType, string dir1, string dir2 = "")
+            public bool BeginInvoke_Called;
+            public Delegate ExpectedDelegate;
+
+            public IAsyncResult BeginInvoke(Delegate method, object[] args)
+            {
+                if (ExpectedDelegate != null)
+                    Assert.Equal(ExpectedDelegate, method);
+
+                BeginInvoke_Called = true;
+                method.DynamicInvoke(args[0], args[1]);
+                return null;
+            }
```
> Are the `TestISynchronizeInvoke` delegates, such as the following, supposed to be invoked?
> 
> https://github.com/dotnet/runtime/blob/6474c725025a8b29b4265c4d9485018982ee6b16/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.cs#L121-L137
> 
> If so, should [`TestISynchronizeInvoke.BeginInvoke()`](https://github.com/dotnet/runtime/blob/6474c725025a8b29b4265c4d9485018982ee6b16/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.cs#L60-L65) be changed to something like the below?
> 
> ```cs
> public IAsyncResult BeginInvoke(Delegate method, object[] args)
> {
>     Assert.Equal(ExpectedDelegate, method);
>     BeginInvoke_Called = true;
>     method.DynamicInvoke(args[0], args[1]);
>     return null;
> }
> ```

I changed this to something like the above, so the following handler would get invoked:

https://github.com/dotnet/runtime/blob/9fa24a29c84ddb49492ca9ce7cfcc9bda82ff264/src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs#L399-L403

thomsj commented:
> @thomsj, all fine questions, but most of them don't seem related to this PR. Are there specific questions related to this fix you still have? Thanks.

As you've noted, I've noticed a few things which aren't directly related to this PR, so my only question is, do you want me to remove anything from this PR, such as the changes to `FileSystemWatcher.OSX.cs`?

stephentoub commented:
> do you want me to remove anything from this PR, such as the changes to FileSystemWatcher.OSX.cs?

Yes, please, thanks.  I'd like to keep this focused on the fix.

thomsj commented:
> > do you want me to remove anything from this PR, such as the changes to FileSystemWatcher.OSX.cs?
> 
> Yes, please, thanks. I'd like to keep this focused on the fix.

I've reverted the `FileSystemWatcher.OSX.cs` changes.

stephentoub commented on src/libraries/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs:
```diff
@@ -34,22 +35,35 @@ public abstract partial class FileSystemWatcherTest : FileCleanupTestBase
+                    {
+                        _output?.WriteLine(ex.ToString());
+                        throw;
+                    }
```
Why is all of this stuff changing?

thomsj replied:
I've reverted the vast majority of the changes to this file.

danmoseley commented:
@stephentoub is next action here on us or @thomsj

stephentoub commented:
> @stephentoub is next action here on us or @thomsj

It's on me.

stephentoub commented:
LGTM.  Thanks, @thomsj.  I rebased your branch on top of main.

thomsj commented:
Thanks a lot for your guidance, @stephentoub.

stephentoub commented:
Thanks for the fix :-)
